File: data-structure/js/arrays/stack-exercise.ts
const readline = require("readline");

const askQuestion = (question: string): Promise<string> => {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise((resolve) => {
    rl.question(question, (response: string) => {
      resolve(response);
      rl.close();
    });
  });
};

class Stack {
  private capacity: number;
  private top: number;
  private values: string[];
  constructor(capacity: number) {
    this.values = new Array(capacity);
    this.capacity = capacity;
    this.top = -1;
  }

  filled() {
    return this.top === this.capacity - 1;
  }

  empty() {
    return this.top === -1;
  }

  stack(value: string) {
    if (this.filled()) console.log("The stack is already filled");
    else {
      this.top++;
      this.values[this.top] = value;
    }
  }

  unstack() {
    if (this.empty()) console.log("The stack is empty");
    else {
      this.top--;
    }
  }

  getTop() {
    if (this.empty()) console.log("The stack is empty");
    else {
      return this.values[this.top];
    }
  }

  validate(text: string) {
    const openers = ["{", "[", "("];
    const closers = ["}", "]", ")"];
    for (let i = 0; i <= text.length; i++) {
      if (openers.includes(text[i])) {
        this.stack(text[i]);
      }
      if (closers.includes(text[i])) {
        switch (text[i]) {
          case "}":
            if (this.getTop() === "{") {
              this.unstack();
              continue;
            }
          case "]":
            if (this.getTop() === "[") {
              this.unstack();
              continue;
            }
          case ")":
            if (this.getTop() === "(") {
              this.unstack();
              continue;
            }
          default:
            throw new Error("Missing enclosing tag");
        }
      }
    }
    console.log("The text is validated ! ;)");
  }
}

askQuestion("Type the text to be validated: ").then((response: string) => {
  const s = new Stack(response.length);
  s.validate(response);
});

File: data-structure/python/arrays/stack-exercise.py
import numpy as np

class Stack:
    def __init__(self, capacity):
        self.__capacity = capacity
        self.__top = -1
        self.__values = np.empty(capacity, dtype=str)
        """ self.__values = np.chararray(self.__capacity, unicode=true) """

    
    def __filled(self):
        if(self.__top == self.__capacity -1):
            return True
        else:
            return False
        
    def __empty(self):
        if(self.__top == -1):
            return True
        else:
            return False
        
    def push(self, value):
        if(self.__filled()):
            print('The stack is already filled')
        else:
            self.__top += 1
            self.__values[self.__top] = value

    def pop(self):
        if(self.__empty()):
            print('The stack is empty')
        else:
            self.__top -= 1

    def top(self):
        if(self.__empty()):
           return -1
        else:
            return self.__values[self.__top]
        
    def validate(self, text):
        openers = ['{', '[', '(']
        closers = ['}', ']', ')']
        try:
            for i in range(len(text)):
                if(text[i] in openers):
                    self.push(text[i])
                if(text[i] in closers):
                    if(text[i] == '}'):
                        if(self.__values[self.__top] == '{'):
                            self.pop()
                            continue
                    if(text[i] == ']'):
                        if(self.__values[self.__top] == '['):
                            self.pop()
                            continue
                    if(text[i] == ')'):
                        if(self.__values[self.__top] == '('):
                            self.pop()
                            continue
                    raise ValueError(f"Valor de fechamento inválido na posição: {i}")
            if(self.__top != -1):
                raise ValueError(f"Tag sem fechamento: {self.__values[self.__top]}")
        except ValueError as e:
            return print(e)
        print('O texto inserido está correto')





    
text = input('Digite o texto para a validação: ')

stack = Stack(len(text))
stack.validate(text)